### PR TITLE
Add PlanFeatures (redesigned plans) to the NUX flow

### PR DESCRIPTION
**Pull request by lamosty**
Fix for #6302. Replaces the current plans implementation at the NUX flow plans page with the redesigned plans components.

### Without Custom Domain

selection_024

### With Custom Domain

selection_022

## Testing Instructions

1. Start Calypso with `ENABLE_FEATURES=manage/plan-features make run`;
2. Navigate to http://calypso.localhost:3000/start/test-plans;
3. Go through the flow with console open;
4. Notice the new redesigned plans page;
5. Does it look alright? Does it look the same as on the `/plans` page?
6. Click on the upgrade btn on each plan;
7. Does it move you to the next step and puts the plan into your cart?
8. Try the above with and without custom domain. It should display four plans without custom domain and three plans with custom domain.

## Todo

- [x] Add Free plan when custom domain is not selected
- [x] Fix visuals of plans header, especially when there are four plans. @gwwar in #6465.

cc @gwwar @artpi @mtias @rralian 

Test live: https://calypso.live/?branch=add/plan-features-nux

### CODE REVIEW

File: client/state/plans/test/selectors.js [hunk 1]
```diff
@@ -113,4 +117,33 @@ describe( 'selectors', () => {
 			expect( price ).to.eql( null );
 		} );
 	} );
+
+	describe( '#getPlanSlug()', () => {
+		it( 'should return plan product_slug', () => {
+			const state = getStateInstance();
```
Comment (gwwar) [inline client/state/plans/test/selectors.js]: Could we maybe avoid using the fixture in these new tests? It makes it easier to read these test cases, even if we end up copying and pasting more code in tests (which is OK).

File: client/my-sites/plans-features-main/index.jsx [hunk 2]
```diff
@@ -50,7 +51,14 @@ class PlansFeaturesMain extends Component {
...
 	}
 
-	getPlanFeatures( site, intervalType ) {
+	getPlanFeatures( params ) {
+		const {
+			site,
+			intervalType,
+			upgradePlan,
+			hideFreePlan
+		} = params;
```
Comment (gwwar) [inline client/my-sites/plans-features-main/index.jsx]: Maybe pull this from `this.props` instead of passing through a params object.

File: client/my-sites/plans-features-main/index.jsx [hunk 3]
```diff
@@ -166,7 +179,15 @@ class PlansFeaturesMain extends Component {
...
 	render() {
-		const { site, plans, isInSignup, sitePlans, intervalType } = this.props;
+		const {
+			site,
+			plans,
+			isInSignup,
+			sitePlans,
+			intervalType,
+			onUpgradeClick,
+			hideFreePlan = false
```
Comment (gwwar) [inline client/my-sites/plans-features-main/index.jsx]: Let's move this to default props since this is a large component.

```javascript
PlansFeaturesMain.defaultProps = {
    hideFreePlan: false
};
```

It'd be nice to also define prop types here too.

File: client/signup/config/flows.js [hunk 4]
```diff
@@ -92,6 +92,13 @@ const flows = {
 		description: 'The current best performing flow in AB tests',
 		lastModified: '2016-05-23'
 	},
+	
+	'plans-features': {
```
Comment (gwwar) [inline client/signup/config/flows.js]: Wow, a lot of flows. 😄  Does anyone know if all of these are active?

File: client/signup/steps/plans-features/index.jsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	plansList() {
+		const {	plans, hideFreePlan } = this.props;
+
+		return (
+			<div>
+				<QueryPlans />
+
+				<PlansFeaturesMain
+					plans={ plans }
+					site={ {} }
```
Comment (gwwar) [inline client/signup/steps/plans-features/index.jsx]: Why do we need to pass through an empty object?

File: client/signup/steps/plans-features/index.jsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+const LocalizedPlansFeaturesStep = localize( PlansFeaturesStep );
+
+export default connect(
+	( state ) => {
+		return {
+			plans: getPlans( state )
+		};
+	}
+)( LocalizedPlansFeaturesStep );
```
Comment (gwwar) [inline client/signup/steps/plans-features/index.jsx]: We can avoid the temporary variable with:

```javascript
export default connect(
//...
)( localize( PlanFeaturesStep ) );
```

File: client/signup/steps/plans-features/index.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+	plansList() {
+		const {	plans, hideFreePlan } = this.props;
+
+		return (
+			<div>
+				<QueryPlans />
+
+				<PlansFeaturesMain
+					plans={ plans }
```
Comment (gwwar) [inline client/signup/steps/plans-features/index.jsx]: We currently pass through `plans` and `sitePlans` for placeholder functionality only. Would it be possible to delegate this logic to `PlanFeatures`?

File: client/state/plans/test/selectors.js [hunk 8]
```diff
@@ -113,4 +117,33 @@ describe( 'selectors', () => {
...
+			const planSlug = getPlanSlug( state, 1337 );
+
+			expect( planSlug ).to.equal( null );
+		} );
+
+		it( 'should return null if plan doesn\'t have product_slug', () => {
+			const state = getStateInstance(),
+				premiumPlanIndex = indexOf( state.plans.items, PLAN_1003 );
+
+			unset( state, `plans.items[${ premiumPlanIndex }].product_slug` );
```
Comment (gwwar) [inline client/state/plans/test/selectors.js]: Let's also avoid the unset

**Comment by gwwar** [comment]
@lamosty Let's update `client/signup/steps/plans/index` to use the feature flag like we do in `client/my-sites/plans/main`. If we do that we won't need to update routes, and we'll more easily get correct abtest results when we wire that in. (We only want the event to fire when a user reaches our new component).

**Comment by gwwar** [comment]
Great work so far in 🔍 all the required pieces!

File: client/state/plans/test/selectors.js (hunk 1 above)
Comment (lamosty) [inline client/state/plans/test/selectors.js]: Alright, agreed. I used fixtures because I saw them and thought that: "if we have those fixtures, why should I duplicate them?". But I see your point.

File: client/my-sites/plans-features-main/index.jsx (hunk 2 above)
Comment (lamosty) [inline client/my-sites/plans-features-main/index.jsx]: Hmm, but why have you not used `this.props` when writing the method initially? Why did you opt to use params instead? I want to know what's different between having two params (and not use `this.props`) and having four (and use `this.props`). :)

File: client/my-sites/plans-features-main/index.jsx (hunk 3 above)
Comment (lamosty) [inline client/my-sites/plans-features-main/index.jsx]: Oh, yes. Thanks for reminder, I almost forgot about this. :)

File: client/signup/steps/plans-features/index.jsx (hunk 5 above)
Comment (lamosty) [inline client/signup/steps/plans-features/index.jsx]: So that using `site.jetpack` does't throw `undefined` error. But we can just specify `site` in `defaultProps` to be an empty obj.

File: client/signup/steps/plans-features/index.jsx (hunk 6 above)
Comment (lamosty) [inline client/signup/steps/plans-features/index.jsx]: Actually, I intentionally wrote it like this. I think this code is more readable and explicit. Don't you think so?

File: client/signup/steps/plans-features/index.jsx (hunk 7 above)
Comment (lamosty) [inline client/signup/steps/plans-features/index.jsx]: Great idea. Will try to implement this.

**Comment by lamosty** [comment]
Thanks! Let me see what can I do about the feature flag in the signup.

File: client/signup/steps/plans-features/index.jsx (hunk 5 above)
Comment (mtias) [inline client/signup/steps/plans-features/index.jsx]: Or just protect against missing `site`.

File: client/signup/steps/plans-features/index.jsx (hunk 6 above)
Comment (gwwar) [inline client/signup/steps/plans-features/index.jsx]: I usually don't separate unwrap these unless I need to export it out for test purposes. Either way is fine, but was just pointing out that this variable wasn't really needed.

File: client/signup/steps/plans-features/index.jsx (hunk 5 above)
Comment (gwwar) [inline client/signup/steps/plans-features/index.jsx]: I think its more common for us to do a falsey check on the object

**Comment by lamosty** [comment]
@gwwar I've implemented all of your suggestions (+ some more). Could you give it additional review please?

File: client/my-sites/plan-features/index.jsx [hunk 9]
```diff
@@ -88,27 +93,31 @@ PlanFeatures.propTypes = {
 	available: PropTypes.bool,
 	onUpgradeClick: PropTypes.func,
 	// either you specify the plan prop or isPlaceholder prop
-	plan: React.PropTypes.string,
-	isPlaceholder: PropTypes.bool
+	plan: PropTypes.string,
+	isPlaceholder: PropTypes.bool,
+	isInSignup: PropTypes.bool
```
Comment (gwwar) [inline client/my-sites/plan-features/index.jsx]: Maybe call this `useSiteSpecificPlans`?

File: client/signup/config/flows.js [hunk 10]
```diff
@@ -87,7 +87,14 @@ const flows = {
 	},
 
 	main: {
-		steps: [ 'survey', 'design-type', 'themes', 'domains', 'plans', 'survey-user' ],
+		steps: [ 
+			'survey',
+			'design-type',
+			'themes',
+			'domains',
+			config.isEnabled( 'manage/plan-features' ) ? 'plansFeatures' : 'plans',
```
Comment (gwwar) [inline client/signup/config/flows.js]: Sorry if I wasn't clear enough earlier, I was suggesting that we modify the `plans` step rather than add `planFeatures` here. This allows up to update the four other flows that use plans. Plus @artpi's shortcut in https://github.com/Automattic/wp-calypso/pull/6451

File: client/my-sites/plan-features/index.jsx (hunk 9 above)
Comment (lamosty) [inline client/my-sites/plan-features/index.jsx]: Sorry, I don't follow you on this one. :) Call `isInSignup` `useSiteSpecificPlans`? If yes, how is that related? What's site specific plans? The `/sites/$siteId/plans` API result?

File: client/signup/config/flows.js (hunk 10 above)
Comment (lamosty) [inline client/signup/config/flows.js]: Oh, sorry, I didn't understand you (and you were specific enough).

We've talked about this internally since and now I see the benefits of doing so.

Thanks. :+1:

File: client/my-sites/plan-features/index.jsx (hunk 9 above)
Comment (gwwar) [inline client/my-sites/plan-features/index.jsx]: It looked like `isInSignup` was being used to determine which data to use, eg from `QuerySitePlans` ( `/sites/$siteId/plans`) or `QueryPlans` (`/plans`). 

A future use case might be that we add another static marketing page, and so the `isInSignup` prop wouldn't make much sense.

Comment (lamosty) [inline client/my-sites/plan-features/index.jsx]: Oh, I see. So, in Calypso UI `my-sites`, that are the site specific plans and in the signup flow that are the general plans.

Okay, nice idea, will do that. :+1:

Comment (gwwar) [inline client/my-sites/plan-features/index.jsx]: The main difference is basically that the site specific plans data, has extra info like discount pricing.

Comment (lamosty) [inline client/my-sites/plan-features/index.jsx]: Good to know, thanks.

File: client/my-sites/plans-features-main/index.jsx (hunk 2 above)
Comment (gwwar) [inline client/my-sites/plans-features-main/index.jsx]: ;) I suppose I should have just pulled it from props before too, or moved the function out of the class if I kept the args.

File: client/signup/steps/plans/index.jsx [hunk 11]
```diff
@@ -138,6 +181,17 @@ module.exports = React.createClass( {
 	},
 
 	render: function() {
+		const showPlanFeatures = config.isEnabled( 'manage/plan-features' );
+
+		if ( showPlanFeatures ) {
+			return (
+				/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Comment (mtias) [inline client/signup/steps/plans/index.jsx]: Please, no :)

File: client/signup/steps/plans/index.jsx [hunk 12]
```diff
@@ -125,6 +144,30 @@ module.exports = React.createClass( {
...
+		const {
+			flowName,
+			stepName,
+			positionInFlow,
+			signupProgressStore
+		} = this.props;
+
+		const translate = this.translate;
+
+		const headerText = translate( 'Pick a plan that\'s right for you.' );
```
Comment (mtias) [inline client/signup/steps/plans/index.jsx]: We prefer double quotes if there's single quotes within the content.

File: client/signup/config/flows.js [hunk 13]
```diff
@@ -87,7 +87,7 @@ const flows = {
 	},
 
 	main: {
-		steps: [ 'survey', 'design-type', 'themes', 'domains', 'plans', 'survey-user' ],
+		steps: [ 'survey', 'design-type', 'themes',	'domains', 'plans', 'survey-user' ],
```
Comment (mtias) [inline client/signup/config/flows.js]: There's a tab here?

File: assets/stylesheets/_components.scss [hunk 14]
```diff
@@ -379,6 +379,7 @@
 @import 'signup/steps/survey/style';
 @import 'signup/steps/design-type/style';
 @import 'signup/steps/plans/style';
+@import 'signup/steps/plans-features/style';
 @import 'signup/steps/paid-plans-only/style';
```
Comment (mtias) [inline assets/stylesheets/_components.scss]: Just spotted this one, should be before alphabetically.

File: client/signup/steps/plans/index.jsx (hunk 11 above)
Comment (lamosty) [inline client/signup/steps/plans/index.jsx]: It's just temporary solution. How would you do it correctly? :)

**Comment by lamosty** [comment]
@gwwar I've refactored the `plans` step to use `plans-features-main` when `manage/plan-features` is enabled. Could you have another look, please?

File: client/signup/steps/plans/index.jsx (hunk 11 above)
Comment (mtias) [inline client/signup/steps/plans/index.jsx]: Fix the class! Temporary solutions are never temporary. :)

Comment (gwwar) [inline client/signup/steps/plans/index.jsx]: @mtias You're welcome to jump in on the branch with the whitespace fixes. I don't think it's really worth the time to fix these up, since devs are highly unlikely to modify the SVGs directly. We'd copy and paste an exported file into these methods instead.

As an aside, I'm not a huge fan of having these SVG's inlined in the component, as they're rather distracting. Right now we have a copy of the SVGs being served in the public folder, and while it would be a separate request, max-age is set appropriately so they'd be cached while you're unproxied. Are we basically worried about the first page load on a cold cache?

Comment (gwwar) [inline client/signup/steps/plans/index.jsx]: @lamosty maybe add a modifier class, if you need something to differentiate this from the original plans step

Comment (gwwar) [inline client/signup/steps/plans/index.jsx]: Eg: 
```
.plans-step.is-wide-layout: {
    margin: auto;
    max-width: 1040px;
}
```

File: client/my-sites/plan-features/index.jsx [hunk 15]
```diff
@@ -117,11 +126,17 @@ export default connect( ( state, ownProps ) => {
...
 		rawPrice: getPlanRawPrice( state, planProductId, showMonthly ),
 		planConstantObj: plansList[ ownProps.plan ],
-		available : canUpgradeToPlan( ownProps.plan ),
-		onUpgradeClick: () => {
-			const selectedSiteSlug = getSiteSlug( state, selectedSiteId );
-			page( `/checkout/${ selectedSiteSlug }/${ getPlanPath( ownProps.plan ) || '' }` );
-		},
+		available: canUpgradeToPlan( ownProps.plan ),
+		onUpgradeClick: ownProps.onUpgradeClick
+			? () => {
```
Comment (gwwar) [inline client/my-sites/plan-features/index.jsx]: In each of the handlers, lets do an early return if `available` is false

Comment (gwwar) [inline client/my-sites/plan-features/index.jsx]: Oh actually on second thought, we only need this in the second handler.

File: client/my-sites/plans-features-main/index.jsx [hunk 16]
```diff
@@ -142,8 +144,13 @@ class PlansFeaturesMain extends Component {
 		);
 	}
 
-	getFAQ( site ) {
-		const { translate } = this.props;
+	getFAQ() {
+		const { site, translate } = this.props;
+
+		if ( this.isJetpackSite( site ) ) {
```
Comment (gwwar) [inline client/my-sites/plans-features-main/index.jsx]: Has a Jetpack site gotten into this method? It looks like we already check for this case below.

File: client/my-sites/plans/main.jsx [hunk 17]
```diff
@@ -177,8 +177,6 @@ const Plans = React.createClass( {
 							showPlanFeatures
 								? <PlansFeaturesMain
 									site={ selectedSite }
-									plans={ this.props.plans }
-									sitePlans={ this.props.sitePlans }
```
Comment (gwwar) [inline client/my-sites/plans/main.jsx]: ✨

File: client/signup/config/steps.js [hunk 18]
```diff
@@ -60,7 +60,7 @@ module.exports = {
 		dependencies: [ 'siteSlug' ],
 		providesDependencies: [ 'cartItem' ]
 	},
-
+
```
Comment (gwwar) [inline client/signup/config/steps.js]: Extra tab?

File: client/signup/steps/plans/index.jsx [hunk 19]
```diff
@@ -125,6 +144,30 @@ module.exports = React.createClass( {
...
+		const {
+			flowName,
+			stepName,
+			positionInFlow,
+			signupProgressStore
+		} = this.props;
+
+		const translate = this.translate;
+
+		const headerText = translate( "Pick a plan that's right for you." );
```
Comment (gwwar) [inline client/signup/steps/plans/index.jsx]: Single quotes please. Did eslint not warn about this one?

**Comment by gwwar** [comment]
👍 I left a few minor notes, but the code looks good. I'll try to test all the major flows today.

File: client/signup/steps/plans/index.jsx (hunk 19 above)
Comment (gwwar) [inline client/signup/steps/plans/index.jsx]: Oh just saw @mtias earlier note. So ignore my note above, and it looks like eslint is properly configured for this too:
```
		'quotes': [ 1, 'single', 'avoid-escape' ],
```

**Comment by gwwar** [comment]
@lamosty let's avoid showing the free plan on the http://calypso.localhost:3000/plans page

**Comment by gwwar** [comment]
screen shot 2016-07-01 at 3 31 58 pm
Looks like I'm unable to select a free plan, on the nux flow.

**Comment by gwwar** [comment]
If we select a free plan, we should also avoid going through checkout.

**Comment by gwwar** [comment]
👍 @lamosty this PR is getting very close!

File: client/signup/steps/plans/index.jsx (hunk 11 above)
Comment (lamosty) [inline client/signup/steps/plans/index.jsx]: Thanks for the suggestion @gwwar.

File: client/signup/steps/plans/style.scss [hunk 20]
```diff
@@ -1,6 +1,11 @@
 .plans-step {
 	margin: 0 auto;
 	max-width: 700px;
+
+	&.is-wide-layout {
```
Comment (mtias) [inline client/signup/steps/plans/style.scss]: When and why is wide layout used?

Comment (lamosty) [inline client/signup/steps/plans/style.scss]: With redesigned plans components. Why: because it doesn't fit and `/plans` page uses it too. I also need to fix `.step-wrapper` because it has `max-width: 960px` only.

File: client/signup/config/steps.js (hunk 18 above)
Comment (lamosty) [inline client/signup/config/steps.js]: Nice catch. Eslint is disabled for that file so I didn't see it.

File: client/signup/steps/plans/style.scss (hunk 20 above)
Comment (mtias) [inline client/signup/steps/plans/style.scss]: So only the redesigned plans need a bigger max-width but the old one will coexist?

Comment (lamosty) [inline client/signup/steps/plans/style.scss]: Yes. Instead of adding a new class name (e.g. `plans-features`), I add `is-wide-layout` if plans features is enabled. This way, I preserve the classnames rule (class name must be the same as folder name).

Do you think this is a bad idea?

**Comment by lamosty** [comment]
@gwwar Thanks for all the feedback; helped much. All the issues fixed!

We will need to rename the upgrade btn on Free plan. It should say `Select Free Plan` instead of `Upgrade`. I think it's a task for another PR because this one is getting huge and you touch the same logic needed for this change in #6465).

**Comment by apeatling** [comment]
I think we need to do some design tweaks here, so let's not merge this until it is done. Right now each plan is much too narrow.

**Comment by gwwar** [comment]
@apeatling Let's fix any visual issues in a follow up PR. This PR is getting large, and it'd be nice to have a full flow available. All new plans redesign work is hidden behind a feature flag so this won't be visible to users.

**Comment by gwwar** [comment]
@lamosty Looks great! Tested the old flows and the new. :shipit:

**Comment by lamosty** [comment]
Thanks for the review @gwwar!

**Comment by melchoyce** [comment]
Should I close out https://github.com/Automattic/wp-calypso/pull/6384 in light of this PR?

**Comment by lamosty** [comment]
@melchoyce hmm, interesting. In your PR, you remove the `max-width` completely. While it's a nice responsive solution, for users with wider monitors, it may cause some issues:

selection_028

The plans on the screenshots above are too wide, not nice and not very readable IMO.

This is how it looks with `max-width: 1040px`. What is nicer to you?

selection_029

If we opt to use the `max-width: 1040px` approach, you can close your PR. If not, let's keep it.

**Comment by melchoyce** [comment]
I think some of the code has changed — when I wrote the PR, there was a 1000px max-width on the container, which was a good width for the four columns. Removing the 700px max-width on the container let it default back to 1000px wide. 1040px is fine :) I was just trying to fix the four columns being smooshed into an artificially small 700px.

**Comment by lamosty** [comment]
Oh, I see. In that case, could you please close your PR? :)

**Comment by melchoyce** [comment]
👍
